File: Hash Table/705. Design HashSet.py
'''
705. Design HashSet
Hash Table, Design

Description:
Design a HashSet without using any built-in hash table libraries.
To be specific, your design should include these functions:
add(value): Insert a value into the HashSet. 
contains(value) : Return whether the value exists in the HashSet or not.
remove(value): Remove a value in the HashSet. If the value does not exist in the HashSet, do nothing.

Example:
MyHashSet hashSet = new MyHashSet();
hashSet.add(1);         
hashSet.add(2);         
hashSet.contains(1);    // returns true
hashSet.contains(3);    // returns false (not found)
hashSet.add(2);          
hashSet.contains(2);    // returns true
hashSet.remove(2);          
hashSet.contains(2);    // returns false (already removed)

Similar Questions:
Design HashMap - Easy
Design Skiplist - Hard
'''

# Solution:
class MyHashSet(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.Set = set()

    def add(self, key):
        """
        :type key: int
        :rtype: None
        """
        return self.Set.add(key)
        

    def remove(self, key):
        """
        :type key: int
        :rtype: None
        """
        if key in self.Set:
            self.Set.remove(key)
        else:
            return False
        

    def contains(self, key):
        """
        Returns true if this set contains the specified element
        :type key: int
        :rtype: bool
        """
        if key in self.Set:
            return True
        else:
            return False
        


# Your MyHashSet object will be instantiated and called as such:
# obj = MyHashSet()
# obj.add(key)
# obj.remove(key)
# param_3 = obj.contains(key)

File: Hash Table/705. Design HashSet.cpp
class MyHashSet {
public:
    /** Initialize your data structure here. */
    MyHashSet() {
        
    }
    
    void add(int key) {
        s.insert(key);
    }
    
    void remove(int key) {
        it = s.find(key);
        if (this->contains(key)) {
            s.erase(key);
        }
    }
    
    /** Returns true if this set contains the specified element */
    bool contains(int key) {
        it = s.find(key);
        if(it != s.end()) {
            return true;
        }else {
            return false;
        }
    }
private:
    set<int> s;
    set<int>::iterator it;
};

/**
 * Your MyHashSet object will be instantiated and called as such:
 * MyHashSet* obj = new MyHashSet();
 * obj->add(key);
 * obj->remove(key);
 * bool param_3 = obj->contains(key);
 */
